pkg/sql/catalog/descriptor.go

```diff
@@ -648,6 +658,11 @@ type TableDescriptor interface {
 	// ones on the table descriptor which are being enforced for all writes, and
 	// "inactive" ones queued in the mutations list.
 	AllActiveAndInactiveUniqueWithoutIndexConstraints() []*descpb.UniqueWithoutIndexConstraint
+	// AllUniqueWithoutIndexConstraints returns all unique constraints that are
+	// not enforced by an index from both the `UniqueWithoutIndex` slice
+	// and mutation slice. It differs from `AllActiveAndInactiveUniqueWithoutIndexConstraints`
+	// in that the return includes UniqueWithoutIndexConstraints that are being dropped.
+	AllUniqueWithoutIndexConstraints() []*descpb.UniqueWithoutIndexConstraint
```

Comment:
I'm more than fine with these constraint-slice functions returning descpb-struct-pointers in this PR, but have you considered introducing a `catalog.Constraint` interface similar to `catalog.Index` or `catalog.Column`, and then constructing these slices in a cache in `tabledesc.wrapper` similar to `indexCache` or `columnCache`?

Doing so would help pay off this technical debt, something which is best done incrementally, and this is a wonderful occasion to do so.